perllib/FixMyStreet/Roles/ParallelAPI.pm

```diff
@@ -2,27 +2,35 @@ package FixMyStreet::Roles::ParallelAPI;
 use Moo::Role;
 use JSON::MaybeXS;
 use Parallel::ForkManager;
-use Storable;
+use Storable qw(retrieve_fd retrieve);
 use Time::HiRes;
+use Digest::MD5 qw(md5_hex);
+use Try::Tiny;
+use Fcntl qw(:flock);
 
 # ---
 # Calling things in parallel
 
 sub call_api {
     # Shifting because the remainder of @_ is passed along further down
-    my ($self, $c, $cobrand, $key) = (shift, shift, shift, shift);
+    my ($self, $c, $cobrand, $key, $fork) = (shift, shift, shift, shift, shift);
 
     my $type = $self->backend_type;
     $key = "$cobrand:$type:$key";
     return $c->session->{$key} if !FixMyStreet->test_mode && $c->session->{$key};
 
-    my $tmp = File::Temp->new;
+    my $calls = encode_json(\@_);
+
+    my $outdir = FixMyStreet->config('WASTEWORKS_BACKEND_TMP_DIR');
+    mkdir($outdir) unless -d $outdir;
+    my $tmp = $outdir . "/" . md5_hex("$key $calls");
```

Comment:
We store the data in the session based solely on the $key, so is there a reason that wouldn't be enough for the tmp file too? Feel I've missed something obvious :)

Reply:
It felt more correct somehow to have the filename based on the actual API calls being made. I think _technically_ it'd be possible for the same `$key` to be used for different batches of calls (e.g. imagine looking up available slots either side of the 15:00 booking cut off) but appreciate it's unlikely. Don't think doing it this way causes any issues though?

Reply:
No, think it's fine.

Reply:
(though if that's the case, would that lead to bad data being stored in the session, or does that not matter). Anyway, is fine if you think it's okay